Throw warnings when zeroing NaNs

valeriabarra opened the pull request:
## Purpose 
The purpose of this PR is to improve the `nans_to_zero` function so that it throws a warning to the user. 

Closes #254 

## Content
- [x] Added docstring
- [x] Added a warning for the user
- [x] Removed a duplicate definition in `experiments/AMIP/moist_mpi_earth/coupler_utils/masker.jl`


Review checklist

I have:
- followed the codebase contribution guide: https://clima.github.io/ClimateMachine.jl/latest/Contributing/
- followed the style guide: https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/
- followed the documentation policy: https://github.com/CliMA/policies/wiki/Documentation-Policy
- checked that this PR does not duplicate an open PR.

In the Content, I have included 
- relevant unit tests, and integration tests, 
- appropriate docstrings on all functions, structs, and modules, and included relevant documentation.

- [x] I have read and checked the items on the review checklist.

valeriabarra commented on src/Regridder.jl:
```diff
@@ -17,9 +17,27 @@ using Dates
+"""
+function nans_to_zero(v)
+    v = isnan(v) ? typeof(v)(0) : v
+    @warn "NaNs were detected and converted to zeros."
```
@charleskawczynski , @LenkaNovak I applied the suggestion in SDI #205 to throw a warning but perhaps it was not a good idea, since it is now flooding the CI logs (see this [build](https://buildkite.com/clima/climacoupler-ci/builds/901)) and we can't even see the entire log as they get truncated because too long.

charleskawczynski replied:
I suggest inlining/deleting this function and adding a warning if any elements are zero.

valeriabarra replied:
What do you mean by inlining/deleting this function? I think those are two different things

charleskawczynski commented on experiments/AMIP/moist_mpi_earth/coupler_utils/masker.jl:
```diff
@@ -70,12 +70,21 @@ function combine_surfaces!(combined_field::Fields.Field, masks::NamedTuple, fiel
+"""
+function nans_to_zero(v)
+    v = isnan(v) ? typeof(v)(0) : v
+    @warn "NaNs were detected and converted to zeros."
```
Based on the way that this function is used, I would recommend not putting a warning here.

valeriabarra replied:
Ok, thanks. So you would put it only in the `Regridder.jl` module

valeriabarra replied:
@charleskawczynski , I have used a global flag. This way, the warning is printed only once at the end of the for loop where this function is called. This is the best solution that I could think of. (Removing the function altogether would have been more difficult because of its vectorized usage). 

It passes CI and is ready for review. Thanks!

charleskawczynski reviewed (commented):
How about we do something like this:

```julia
#= Converts NaNs to zeros of the same type. =#
nans_to_zero(v) = isnan(v) ? typeof(v)(0) : v

function combine_surfaces!(combined_field::Fields.Field, masks::NamedTuple, fields::NamedTuple)
    combined_field .= eltype(combined_field)(0)
    warn_nans = false
    for surface_name in propertynames(fields) # could use dot here?
        if any(x->isnan, getproperty(fields, surface_name))
            warn_nans = true
        end
        @. combined_field += getproperty(masks, surface_name) * nans_to_zero.(getproperty(fields, surface_name))
    end
    warn_nans && @warn "NaNs were detected and converted to zeros."
end
```

I've converted the doc string to a comment, since this is not really a user-facing function.

valeriabarra commented:
> A module variable is not going to be good w.r.t. inference, so we should nix that.
> 
> How about we do something like this:
> 
> ```julia
> #= Converts NaNs to zeros of the same type. =#
> nans_to_zero(v) = isnan(v) ? typeof(v)(0) : v
> 
> function combine_surfaces!(combined_field::Fields.Field, masks::NamedTuple, fields::NamedTuple)
>     combined_field .= eltype(combined_field)(0)
>     warn_nans = false
>     for surface_name in propertynames(fields) # could use dot here?
>         if any(x->isnan, getproperty(fields, surface_name))
>             warn_nans = true
>         end
>         @. combined_field += getproperty(masks, surface_name) * nans_to_zero.(getproperty(fields, surface_name))
>     end
>     warn_nans && @warn "NaNs were detected and converted to zeros."
> end
> ```
> 
> I've converted the doc string to a comment, since this is not really a user-facing function.

Thank you. Just to let you know, it doesn't quite work the way it is because of `TypeError: non-boolean (typeof(isnan)) used in boolean context`.

charleskawczynski commented:
Right, need `any(x->isnan(x), getproperty(fields, surface_name))`

charleskawczynski commented:
Instead of
```julia
@. combined_field += getproperty(masks, surface_name) * nans_to_zero.(getproperty(fields, surface_name))
```
you need
```julia
combined_field .+= getproperty(masks, surface_name) .* nans_to_zero.(getproperty(fields, surface_name))
```

charleskawczynski reviewed (approved):
Nice, it looks like this also reduces allocations from 2725088 to 2685440.

If you want, we could also address the TODO in `perf/flame.jl`:

`[ Info: TODO: lower allocs_limit[perf_default_modular] to: 2725088`

I'm fine with that going in a separate PR, though.

valeriabarra commented:
> 2685440

Thanks Charlie! I have updated the allocation limits for all of the 3: `perf_default_modular`, `perf_coarse_single_modular`, `perf_target_amip_n32_shortrun`

charleskawczynski commented:
Feel free to merge / squash & merge

valeriabarra commented:
bors r+
